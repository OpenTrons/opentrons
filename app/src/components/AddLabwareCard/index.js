// @flow
import * as React from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { Card } from '@opentrons/components'

import { getConfig } from '../../config'
import {
  changeCustomLabwareDirectory,
  addCustomLabware,
  clearAddCustomLabwareFailure,
  getAddLabwareFailure,
} from '../../custom-labware'

import { ManagePath } from './ManagePath'
import { AddLabware } from './AddLabware'
import { PortaledAddLabwareFailureModal } from './AddLabwareFailureModal'

import type { Dispatch } from '../../types'

// TODO(mc, 2019-10-17): i18n
const LABWARE_MANAGEMENT = 'Labware Management'

export function AddLabwareCard() {
  const dispatch = useDispatch<Dispatch>()
  const config = useSelector(getConfig)
  const addFailure = useSelector(getAddLabwareFailure)
  const labwarePath = config.labware.directory
  const handleChangePath = () => dispatch(changeCustomLabwareDirectory())
  const handleAddLabware = () => dispatch(addCustomLabware())
  const showAddFailure = addFailure.file || addFailure.errorMessage !== null

  return (
    <Card title={LABWARE_MANAGEMENT}>
      <ManagePath path={labwarePath} onChangePath={handleChangePath} />
      <AddLabware onAddLabware={handleAddLabware} />
      {showAddFailure && (
        <PortaledAddLabwareFailureModal
          {...addFailure}
          directory={labwarePath}
          onCancel={() => dispatch(clearAddCustomLabwareFailure())}
          onOverwrite={file => dispatch(addCustomLabware(file))}
        />
      )}
    </Card>
  )
}
