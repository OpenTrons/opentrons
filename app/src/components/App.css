/* TODO(mc, 2018-02-09): fix lint errors and remove this ignore */
/* stylelint-disable font-family-no-missing-generic-family-keyword  */
* {
  margin: 0;
  padding: 0;
  box-sizing: border-box;
}

.wrapper {
  width: 1024px;
  height: 100vh;
  display: flex;
  flex-direction: row;
  font-family: 'sans';
  position: fixed;
}
