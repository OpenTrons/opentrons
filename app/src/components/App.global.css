@import '@opentrons/components';

* {
  box-sizing: border-box;
}

html,
body {
  width: 100%;
  height: 100%;
  margin: 0;
  padding: 0;
  color: var(--c-font-dark);
}

a {
  text-decoration: none;
}

.fixed {
  position: fixed;
}

.relative {
  position: relative;
}

.absolute {
  position: absolute;
}

.absolute__fill {
  left: 0;
  right: 0;
  top: 0;
  bottom: 0;
}

.inline {
  display: inline;
}

.block {
  display: block;
}

.inline_block {
  display: inline-block;
}

/*** TYPOGRAPHY ***/
@font-face {
  font-family: 'trons';
  src: url('../fonts/Geogtq-Rg-webfont.woff2');
  font-weight: normal;
  font-style: normal;
}

@font-face {
  font-family: 'sans';
  src: url('../fonts/OpenSans-Regular-webfont.woff');
  font-weight: normal;
  font-style: normal;
}

.sans {
  font-family: sans;
}

.trons {
  font-family: trons;
}

/*** BUTTONS ***/
button {
  border: none;
}

button:focus {
  outline: 0;
}

.btn {
  height: 2.5rem;
  line-height: 2.5rem;
  border-radius: 2px;
  font-family: 'sans';
  font-size: 14px;
  text-transform: uppercase;
  cursor: pointer;
  background-color: transparent;
}

.btn:active {
  outline: none;
  border: none;
}

.disabled {
  pointer-events: none;
  opacity: 0.5;
}

.btn_light {
  border: 1px solid silver;
  background-color: #fff;
  color: #666;
}

.btn_light:hover {
  background-color: #eee;
}

.btn_dark {
  background-color: #9b9b9b;
  color: white;
}

.btn_dark:hover {
  background-color: #777;
}

.btn_blue {
  background-color: #006fff;
  color: white;
}

.btn_blue:hover {
  background-color: #00f;
}
