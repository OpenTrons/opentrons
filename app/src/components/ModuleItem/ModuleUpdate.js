// @flow
import * as React from 'react'
import last from 'lodash/last'
import { useSelector, useDispatch } from 'react-redux'
import {
  OutlineButton,
  HoverTooltip,
  Icon,
  AlertModal,
} from '@opentrons/components'
import { Portal } from '../portal'
import { getConnectedRobotName } from '../../robot/selectors'
import {
  useDispatchApiRequest,
  getRequestById,
  PENDING,
  FAILURE,
  dismissRequest,
} from '../../robot-api'
import { updateModule } from '../../modules'
import type { UpdateModuleAction } from '../../modules/types'
import type { State, Dispatch } from '../../types'
import type { RequestState } from '../../robot-api/types'
import styles from './styles.css'

const FW_IS_UP_TO_DATE = 'Module Firmware is up to date'
const CONNECT_TO_UPDATE = 'Connect to Robot to update Module'
const OK_TEXT = 'Ok'

const UPDATE = 'update'
const UPDATE_TO_DATE = 'up to date'

const FAILED_UPDATE_HEADER = 'Failed to update Module Firmware'
const FAILED_UPDATE_BODY =
  'An error occurred while attempting to update your robot.'

type Props = {|
  hasAvailableUpdate: boolean,
  canControl: boolean,
  moduleId: string,
|}

export function ModuleUpdate(props: Props) {
  const { hasAvailableUpdate, moduleId, canControl } = props
  const dispatch = useDispatch<Dispatch>()
  const robotName = useSelector(getConnectedRobotName)
  const [
    dispatchApiRequest,
    requestIds,
  ] = useDispatchApiRequest<UpdateModuleAction>()

  const handleClick = () => {
    canControl &&
      robotName &&
      dispatchApiRequest(updateModule(robotName, moduleId))
  }
  const latestRequestId = last(requestIds)
  const latestRequest = useSelector<State, RequestState | null>(state =>
    getRequestById(state, latestRequestId)
  )
  const isPending = latestRequest?.status === PENDING

  const buttonText = hasAvailableUpdate ? UPDATE : UPDATE_TO_DATE
  let tooltipText = null
  if (!canControl) tooltipText = CONNECT_TO_UPDATE
  if (!hasAvailableUpdate) tooltipText = FW_IS_UP_TO_DATE

  const handleCloseErrorModal = () => {
    dispatch(dismissRequest(latestRequestId))
  }
  return (
    <div className={styles.module_update_wrapper}>
      <HoverTooltip tooltipComponent={tooltipText}>
        {hoverTooltipHandlers => (
          <div {...hoverTooltipHandlers}>
            <OutlineButton
              className={styles.module_update_button}
              onClick={handleClick}
              disabled={!canControl || !hasAvailableUpdate || isPending}
            >
              {isPending ? (
                <Icon name="ot-spinner" height="1em" spin />
              ) : (
                buttonText
              )}
            </OutlineButton>
          </div>
        )}
      </HoverTooltip>
      {latestRequest && latestRequest.status === FAILURE && (
        <Portal>
          <AlertModal
            alertOverlay
            heading={FAILED_UPDATE_HEADER}
            buttons={[{ children: OK_TEXT, onClick: handleCloseErrorModal }]}
          >
            <p>{FAILED_UPDATE_BODY}</p>
            <p>{latestRequest.error.message}</p>
          </AlertModal>
        </Portal>
      )}
    </div>
  )
}
