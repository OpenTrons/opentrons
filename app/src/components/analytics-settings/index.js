// @flow
import AnalyticsSettingsCard from './AnalyticsSettingsCard'
import AnalyticsSettingsModal from './AnalyticsSettingsModal'

export { AnalyticsSettingsCard, AnalyticsSettingsModal }
