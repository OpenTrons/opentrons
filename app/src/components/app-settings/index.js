// @flow
// app settings related components
export * from './UpdateAppModal'
export * from './AppSoftwareSettingsCard'
export * from './AppAdvancedSettingsCard'
