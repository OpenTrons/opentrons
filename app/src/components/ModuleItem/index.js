// @flow
import * as React from 'react'

import type {Module} from '../../http-api-client'

import ModuleImage from './ModuleImage'
import ModuleInfo from './ModuleInfo'
import ModuleUpdate from './ModuleUpdate'
import NoModulesMessage from './NoModulesMessage'

import styles from './styles.css'

type Props = Module & {
  availableUpdate?: ?string
}

export default function ModuleItem (props: Props) {
  return (
    <div className={styles.module_item}>
      <ModuleImage name={props.name}/>
      <ModuleInfo {...props}/>
      <ModuleUpdate availableUpdate={props.availableUpdate}/>
    </div>
  )
}

export {NoModulesMessage}
