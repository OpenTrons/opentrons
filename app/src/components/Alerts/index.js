// @flow
import * as React from 'react'
import { useSelector, useDispatch } from 'react-redux'
import head from 'lodash/head'

import * as AppAlerts from '../../alerts'
import { LostConnectionAlert } from '../LostConnectionAlert'
import { AnalyticsSettingsModal } from '../analytics-settings'
import { U2EDriverOutdatedAlert } from './U2EDriverOutdatedAlert'

import type { State, Dispatch } from '../../types'
import type { AlertId } from '../../alerts/types'

export function Alerts(): React.Node {
  const dispatch = useDispatch<Dispatch>()

  // TODO(mc, 2020-05-07): move head logic to selector with alert priorities
  const activeAlert: AlertId | null = useSelector((state: State) => {
    return head(AppAlerts.getActiveAlerts(state)) ?? null
  })

  const dismissAlert = remember => {
    if (activeAlert) {
      dispatch(AppAlerts.alertDismissed(activeAlert, remember))
    }
  }

  return (
    <>
      {/* TODO(mc, 2020-05-07): LostConnectionAlert currently controls its own
          render; move its logic into `state.alerts` */}
      <LostConnectionAlert />
      {/* TODO(mc, 2020-05-07): AnalyticsSettingsModal currently controls its
          own render; move its logic into `state.alerts` */}
      <AnalyticsSettingsModal />

      {activeAlert === AppAlerts.ALERT_U2E_DRIVER_OUTDATED ? (
        <U2EDriverOutdatedAlert dismissAlert={dismissAlert} />
      ) : null}
    </>
  )
}
