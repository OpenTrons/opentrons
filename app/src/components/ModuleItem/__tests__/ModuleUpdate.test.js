// @flow
import * as React from 'react'
import { Provider } from 'react-redux'
import { shallow, mount } from 'enzyme'
import configureMockStore from 'redux-mock-store'

import { Button } from '@opentrons/components'
import { ModuleUpdate } from '../ModuleUpdate'

const mockStore = configureMockStore([])

describe('ModuleUpdate', () => {
  let store
  beforeEach(() => {
    store = mockStore({
      mockState: true,
      robot: {
        connection: { connectedTo: 'BITTER-MOCK' },
      },
      robotApi: {},
    })
  })

  afterEach(() => {
    jest.resetAllMocks()
  })

  test('component renders', () => {
    const treeAvailableCanControl = shallow(
      <Provider store={store}>
        <ModuleUpdate
          canControl={true}
          moduleId="FAKEMODULE1234"
          hasAvailableUpdate={true}
        />
      </Provider>
    )
    const treeAvailableNoControl = shallow(
      <Provider store={store}>
        <ModuleUpdate
          canControl={false}
          moduleId="FAKEMODULE1234"
          hasAvailableUpdate={true}
        />
      </Provider>
    )
    const treeNotAvailableCanControl = shallow(
      <Provider store={store}>
        <ModuleUpdate
          canControl={true}
          moduleId="FAKEMODULE1234"
          hasAvailableUpdate={false}
        />
      </Provider>
    )
    const treeNotAvailableNoControl = shallow(
      <Provider store={store}>
        <ModuleUpdate
          canControl={false}
          moduleId="FAKEMODULE1234"
          hasAvailableUpdate={false}
        />
      </Provider>
    )
    expect(treeAvailableCanControl).toMatchSnapshot()
    expect(treeNotAvailableCanControl).toMatchSnapshot()
    expect(treeAvailableNoControl).toMatchSnapshot()
    expect(treeNotAvailableNoControl).toMatchSnapshot()
  })

  test('displays a Warning for invalid files', () => {
    const SPECS = [
      {
        canControl: true,
        hasAvailableUpdate: true,
        expectDisabled: false,
      },
      {
        canControl: true,
        hasAvailableUpdate: false,
        expectDisabled: true,
      },
      {
        canControl: false,
        hasAvailableUpdate: true,
        expectDisabled: true,
      },
      {
        canControl: false,
        hasAvailableUpdate: false,
        expectDisabled: true,
      },
    ]

    SPECS.forEach(({ canControl, hasAvailableUpdate, expectDisabled }) => {
      const wrapper = mount(
        <Provider store={store}>
          <ModuleUpdate
            moduleId="FAKEMODULE1234"
            canControl={canControl}
            hasAvailableUpdate={hasAvailableUpdate}
          />
        </Provider>
      )
      expect(wrapper.find(Button).prop('disabled')).toEqual(expectDisabled)
    })
  })
})
