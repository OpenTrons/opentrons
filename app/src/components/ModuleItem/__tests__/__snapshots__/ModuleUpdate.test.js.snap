// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModuleUpdate component renders 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "clear": [Function],
          "get": [Function],
          "notify": [Function],
          "subscribe": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": [Function],
      },
    }
  }
>
  <ModuleUpdate
    canControl={true}
    hasAvailableUpdate={true}
    moduleId="FAKEMODULE1234"
  />
</ContextProvider>
`;

exports[`ModuleUpdate component renders 2`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "clear": [Function],
          "get": [Function],
          "notify": [Function],
          "subscribe": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": [Function],
      },
    }
  }
>
  <ModuleUpdate
    canControl={true}
    hasAvailableUpdate={false}
    moduleId="FAKEMODULE1234"
  />
</ContextProvider>
`;

exports[`ModuleUpdate component renders 3`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "clear": [Function],
          "get": [Function],
          "notify": [Function],
          "subscribe": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": [Function],
      },
    }
  }
>
  <ModuleUpdate
    canControl={false}
    hasAvailableUpdate={true}
    moduleId="FAKEMODULE1234"
  />
</ContextProvider>
`;

exports[`ModuleUpdate component renders 4`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "clear": [Function],
          "get": [Function],
          "notify": [Function],
          "subscribe": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": [Function],
      },
    }
  }
>
  <ModuleUpdate
    canControl={false}
    hasAvailableUpdate={false}
    moduleId="FAKEMODULE1234"
  />
</ContextProvider>
`;
